// git-parse/dist/parse_commit.js
Object.defineProperty(exports, "__esModule", {
  value: true
});

var _git_log_format_markers = require('./constants/git_log_format_markers');

var parseCommit = function parseCommit(commit) {
  var hash = commit[0];
  var authorName = commit[1];
  var authorEmail = commit[2];
  var date = commit[3];

  var messageIndex = commit.findIndex(function (line) {
    return line.match(_git_log_format_markers.gitLogMessageMarker);
  });
  var fileIndex = commit.findIndex(function (line) {
    return line.match(_git_log_format_markers.gitLogFileMarker);
  });
  var message = commit.slice(messageIndex + 1, fileIndex).join('\n');
  var files = commit.slice(fileIndex + 1);

  var addPattern = /^A\s([^\s]+)/;
  var deletePattern = /^D\s([^\s]+)/;
  var modifyPattern = /^M\s([^\s]+)/;
  var renamePattern = /^R[0-9]+\s(.+)\s(.+)/;

  var filterFileChanges = function filterFileChanges(pattern) {
    return files.reduce(function (accumulator, file) {
      var match = file.match(pattern);
      if (match) {
        accumulator.push({ path: match[1] });
      }

      return accumulator;
    }, []);
  };

  var filesRenamed = files.reduce(function (accumulator, file) {
    var match = file.match(renamePattern);
    if (match) {
      accumulator.push({
        oldPath: match[1],
        newPath: match[2]
      });
    }
    return accumulator;
  }, []);

  var parsedCommit = {
    hash: hash,
    authorName: authorName,
    authorEmail: authorEmail,
    date: date,
    message: message,
    filesAdded: filterFileChanges(addPattern),
    filesDeleted: filterFileChanges(deletePattern),
    filesModified: filterFileChanges(modifyPattern),
    filesRenamed: filesRenamed
  };

  return parsedCommit;
};

exports.default = parseCommit;
